### PR TITLE
Fixing bug in csisnapshot

### PR DESCRIPTION
**Pull request by jainsmit**
**What this PR does / why we need it**:
1. Fixing the issue with Csi Snapshot and restore code.
2. Added code for wait for snapshot
3. Added code to check DM path in volume inspect
4. Added code to check pure file volume
5. Added code to not trigger clones for FB volumes

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

http://jen-endurance.pwx.dev.purestorage.com/job/Torpedo/job/tp-fada-longevity/67/console
https://jenkins.pwx.dev.purestorage.com/job/Dev/job/torpedo-dev03/36/console

### CODE REVIEW

File: drivers/volume/portworx/portworx.go [hunk 1]
```diff
@@ -1992,6 +2006,35 @@ func (d *portworx) IsStorageExpansionEnabled() (bool, error) {
 
 // IsPureVolume return true if volume is pure volume else false
 func (d *portworx) IsPureVolume(volume *torpedovolume.Volume) (bool, error) {
+	var proxySpec *api.ProxySpec
+	var err error
+	if proxySpec, err = d.getProxySpecForAVolume(volume); err != nil {
+		return false, err
+	}
+	if proxySpec == nil {
```
Comment (Rohit-PX) [inline drivers/volume/portworx/portworx.go:2014]: This check is present twice.

File: drivers/volume/portworx/portworx.go [hunk 2]
```diff
@@ -1992,6 +2006,35 @@ func (d *portworx) IsStorageExpansionEnabled() (bool, error) {
...
+
+	if proxySpec == nil {
+		return false, nil
+	}
+	if proxySpec.ProxyProtocol == api.ProxyProtocol_PROXY_PROTOCOL_PURE_BLOCK {
+		logrus.Debugf("Volume is Pure Block volume: %s", volume.ID)
+		return true, nil
+	}
+
+	if proxySpec.ProxyProtocol == api.ProxyProtocol_PROXY_PROTOCOL_PURE_FILE {
```
Comment (Rohit-PX) [inline drivers/volume/portworx/portworx.go:2026]: If condition in line 2021 is true, this line will never be executed as the function would return from line 2021.

If we need to check only one of the conditions, a simpler/better way would be:

```
if proxySpec.ProxyProtocol == api.ProxyProtocol_PROXY_PROTOCOL_PURE_BLOCK  || proxySpec.ProxyProtocol == api.ProxyProtocol_PROXY_PROTOCOL_PURE_FILE {
```

File: tests/testTriggers.go [hunk 3]
```diff
@@ -1240,6 +1259,22 @@ func TriggerVolumeClone(contexts *[]*scheduler.Context, recordChan *chan *EventR
 			})
 			for _, vol := range appVolumes {
 				var clonedVolID string
+
+				// Skip clone trigger for Pure FB volumes
+				isPureFileVol, err := Inst().V.IsPureFileVolume(vol)
+				if err != nil {
+					logrus.Errorf("Checking pure file volume is failing with error: %v", err)
```
Comment (Rohit-PX) [inline tests/testTriggers.go:1266]: It will be helpful to print the volume name/ID along with the error here.

File: tests/testTriggers.go [hunk 4]
```diff
@@ -4355,6 +4414,33 @@ func rangeStructer(args ...interface{}) []interface{} {
...
+		ObjectMeta:        v1obj,
+		Provisioner:       k8s.CsiProvisioner,
+		Parameters:        params,
+		ReclaimPolicy:     &reclaimPolicyDelete,
+		VolumeBindingMode: &bindMode,
+	}
+
+	sc, err := k8sStorage.CreateStorageClass(&scObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CsiSnapshot storage class: %s", name)
```
Comment (Rohit-PX) [inline tests/testTriggers.go:4439]:
```suggestion
		return nil, fmt.Errorf("failed to create CsiSnapshot storage class: %s. Error: %v", name, err)
```

File: drivers/scheduler/k8s/k8s.go [hunk 5]
```diff
@@ -4898,21 +4918,34 @@ func (k *K8s) RecycleNode(n node.Node) error {
...
 	for _, specObj := range ctx.App.SpecList {
+
 		if obj, ok := specObj.(*corev1.PersistentVolumeClaim); ok {
 			pvc, _ := k8sCore.GetPersistentVolumeClaim(obj.Name, obj.Namespace)
-			snapshotOkay, err := k.filterPureVolumesIfEnabled(pvc)
+			snapshotOkay, err := k.filterPureTypeVolumeIfEnabled(pvc, volTypes)
 			if err != nil {
 				return nil, err
 			}
 			if snapshotOkay {
```
Comment (Rohit-PX) [inline drivers/scheduler/k8s/k8s.go:4934]: add and else for what happens if `snapshotOkay` is false

**Review by Rohit-PX — CHANGES_REQUESTED** [review]
Looks good except the minor changes. Also post the link to test torpedo runs.

File: tests/testTriggers.go (hunk 4 above)
Comment (jainsmit) [reply tests/testTriggers.go]: Incorporated !!

File: tests/testTriggers.go (hunk 3 above)
Comment (jainsmit) [reply tests/testTriggers.go]: Incorporated !!

**Review by Rohit-PX — APPROVED** [review]

**Review by Pure-AdamuKaapan — APPROVED** [review]
LGTM! Cool to see more Pure tests making it in

**Pull request merged**